src/libcore/mem.rs

```diff
@@ -1039,6 +1039,19 @@ impl<T> MaybeUninit<T> {
     /// Note that dropping a `MaybeUninit` will never call `T`'s drop code.
     /// It is your responsibility to make sure `T` gets dropped if it got initialized.
     #[unstable(feature = "maybe_uninit", issue = "53491")]
+    #[rustc_const_unstable(feature = "const_maybe_uninit_zeroed")]
+    #[cfg(not(stage0))]
+    pub const fn zeroed() -> MaybeUninit<T> {
+        let mut u = MaybeUninit::<T>::uninitialized();
+        unsafe {
+            (&mut *u.value as *mut T).write_bytes(0u8, 1);
```

Comment:
Why not make `as_mut_ptr` a `const fn` as well and avoid duplicating?

Reply:
I've been digging into making that change, but I keep running into issues [like this](https://play.rust-lang.org/?gist=512c5b2f6b97882eaa8d465d58dd2be5&version=nightly&mode=debug&edition=2018), and now that I think of it, I can't figure out why this even compiles in the first place:

- `ManuallyDrop` implements the deref traits, but those functions aren't `const`.
- I only made the intrinsic `write_bytes` `const`, not the [`write_bytes` method for `mut_ptr` lang item](https://github.com/rust-lang/rust/blob/423d8109868c1f926f2cfcc3bff980c3daa515fd/src/libcore/ptr.rs#L2223-L2229). This should be using the latter, not the intrinsic.
- This requires taking a mutable reference.

I don't understand why this even works as-is. Trying to make `as_mut_ptr` a `const fn` reveals all these issues, but manually inlining it here does not...

Reply:
O_o, that looks like a hole in the const checks, but I don't even know where to start debugging this. Even if the function isn't checked for `min_const_fn`, it should definitely fail to build due to the regular const fn checks.

Reply:
Yea, in the playground this won't compile: https://play.rust-lang.org/?gist=adce30a361142dcea174aee030ea8a91&version=nightly&mode=debug&edition=2018

Reply:
Could this be related to what https://github.com/rust-lang/rust/pull/54715 fixes?

Reply:
No, I think I found it. `is_min_const_fn` doesn't check for `is_const_fn` as a prerequisite.

The fix is to add `self.is_const_fn(def_id) &&` infront of https://github.com/rust-lang/rust/blob/567557f630693d47fd21151ff1fdbc430e330a13/src/librustc/ty/constness.rs#L43